File: scripts/instagram-upload.py
#!/usr/bin/env python3
"""
Instagram upload script via Composio Python SDK.
Called from Node.js API route as a child process.

Usage:
    python instagram-upload.py <json_args>
    Args JSON: {"entity_id":"shohib","ig_user_id":"...","video_url":"...","caption":"...","share_to_feed":true}
"""

import json
import os
import sys
import traceback


def main():
    try:
        if len(sys.argv) < 2:
            print(json.dumps({"success": False, "error": "No arguments provided"}))
            sys.exit(1)

        args = json.loads(sys.argv[1])
        entity_id = args.get("entity_id", "shohib")
        ig_user_id = args.get("ig_user_id", "")
        video_url = args.get("video_url", "")
        caption = args.get("caption", "")
        share_to_feed = args.get("share_to_feed", True)

        if not video_url:
            print(json.dumps({"success": False, "error": "video_url is required"}))
            sys.exit(1)

        if not ig_user_id:
            print(json.dumps({"success": False, "error": "ig_user_id is required"}))
            sys.exit(1)

        api_key = os.environ.get("COMPOSIO_API_KEY", "")
        if not api_key:
            print(json.dumps({"success": False, "error": "COMPOSIO_API_KEY is required"}))
            sys.exit(1)

        os.environ["COMPOSIO_API_KEY"] = api_key

        from composio import Composio

        composio = Composio(toolkit_versions={"INSTAGRAM": "20260501_00"})

        # Step 1: Create media container
        create_result = composio.tools.execute(
            "INSTAGRAM_POST_IG_USER_MEDIA",
            user_id=entity_id,
            arguments={
                "ig_user_id": ig_user_id,
                "video_url": video_url,
                "media_type": "REELS",
                "caption": caption,
                "share_to_feed": share_to_feed,
            },
        )

        if not create_result.get("successful"):
            error_msg = create_result.get("error") or "Failed to create Instagram media container"
            print(json.dumps({"success": False, "error": error_msg}))
            sys.exit(1)

        creation_id = create_result.get("data", {}).get("id")
        if not creation_id:
            print(json.dumps({"success": False, "error": "No container ID returned"}))
            sys.exit(1)

        # Step 2: Publish the container (with auto-wait for video processing)
        publish_result = composio.tools.execute(
            "INSTAGRAM_POST_IG_USER_MEDIA_PUBLISH",
            user_id=entity_id,
            arguments={
                "ig_user_id": ig_user_id,
                "creation_id": creation_id,
                "max_wait_seconds": 300,
            },
        )

        if not publish_result.get("successful"):
            error_msg = publish_result.get("error") or "Failed to publish Instagram media"
            print(json.dumps({"success": False, "error": error_msg, "container_id": creation_id}))
            sys.exit(1)

        media_id = publish_result.get("data", {}).get("id", "")
        print(json.dumps({
            "success": True,
            "container_id": creation_id,
            "media_id": media_id,
        }))

    except Exception as e:
        tb = traceback.format_exc()
        print(json.dumps({"success": False, "error": str(e), "traceback": tb[-500:]}))
        sys.exit(1)


if __name__ == "__main__":
    main()

File: src/lib/composio/instagram-upload.ts
export type InstagramUploadOptions = {
  entityId: string;
  igUserId: string;
  videoUrl: string;
  caption?: string;
  shareToFeed?: boolean;
};

export type InstagramUploadResult = {
  success: boolean;
  containerId?: string;
  mediaId?: string;
  error?: string;
};

export async function uploadInstagramReels(
  options: InstagramUploadOptions,
): Promise<InstagramUploadResult> {
  const {
    entityId,
    igUserId,
    videoUrl,
    caption = "",
    shareToFeed = true,
  } = options;

  const apiKey = process.env.COMPOSIO_API_KEY;

  if (!apiKey) {
    return { success: false, error: "COMPOSIO_API_KEY is not configured" };
  }

  try {
    // Step 1: Create media container
    let creationId = "";

    const createRes = await fetch(
      "https://backend.composio.dev/api/v2/tools/INSTAGRAM_POST_IG_USER_MEDIA/execute",
      {
        method: "POST",
        headers: {
          Authorization: `Bearer ${apiKey}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          ig_user_id: igUserId,
          video_url: videoUrl,
          media_type: "REELS",
          caption,
          share_to_feed: shareToFeed,
        }),
      },
    );

    if (!createRes.ok) {
      const text = await createRes.text();
      return {
        success: false,
        error: `Failed to create media container (${createRes.status}): ${text.slice(0, 200)}`,
      };
    }

    const createData = await createRes.json();
    creationId = createData?.data?.id ?? createData?.id ?? "";

    if (!creationId) {
      return {
        success: false,
        error: "No container ID returned from Instagram media creation",
      };
    }

    // Step 2: Publish the container
    const publishRes = await fetch(
      "https://backend.composio.dev/api/v2/tools/INSTAGRAM_POST_IG_USER_MEDIA_PUBLISH/execute",
      {
        method: "POST",
        headers: {
          Authorization: `Bearer ${apiKey}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          ig_user_id: igUserId,
          creation_id: creationId,
          max_wait_seconds: 300,
        }),
      },
    );

    if (!publishRes.ok) {
      const text = await publishRes.text();
      return {
        success: false,
        error: `Failed to publish media (${publishRes.status}): ${text.slice(0, 200)}`,
        containerId: creationId,
      };
    }

    const publishData = await publishRes.json();
    const mediaId = publishData?.data?.id ?? "";

    return {
      success: true,
      containerId: creationId,
      mediaId,
    };
  } catch (error) {
    return {
      success: false,
      error:
        error instanceof Error
          ? error.message
          : "Unknown error during Instagram upload",
    };
  }
}
